feat: callout macro

sdelamo opened the pull request:
This aims to remove duplication of copy between guides. I think it is key to explain what Micronaut annotations do, but it is key to keep guides easy to update.

sdelamo committed:
feat: callout macro

ilopmar reviewed (commented):
I'm not really sure about this change. One nice thing about the new infrastructure is that all the guide is in the same file.

With this change we now have +20 new text snippets that only contain one line. It makes more complicated to write a guide because we need to review all those snippets and also it makes more complicated to review new PRs because we always need to go to the snippets.

For me this is a :-1:

sdelamo commented:
> I'm not really sure about this change. One nice thing about the new infrastructure is that all the guide is in the same file.
> 
> With this change we now have +20 new text snippets that only contain one line. It makes more complicated to write a guide because we need to review all those snippets and also it makes more complicated to review new PRs because we always need to go to the snippets.
> 
> For me this is a 👎

It is just common snippets but even faster. You don't need to use them when you start writing the guide if you don't want. I have been editing most guides and it is fast. You want to do the callout about `Singleton` `callout:singleton[1]`. Etc.  It does not make sense that you have `@Introspected` copy duplicated in twenty files. 

From the perspective of someone reviewing the guide. It is the same as common snippets.

burtbeckwith reviewed (approved):
It is a bit weird to have all those 1-line includes, but I like that it improves consistency. I tend to look at other guides to copy/paste from for things like these, so in the future I'd just copy the macro instead of the full text, or neither and write my own inconsistent version.

One thing that would help would be a list of these somewhere. I've been thinking about adding some more docs for guide authors, mostly things I discovered only by reading lots of code or in PR comments. I added https://github.com/micronaut-projects/micronaut-guides/issues/420 to get this into the queue

ilopmar commented:
> It is just common snippets but even faster. You don't need to use them when you start writing the guide if you don't want. I have been editing most guides and it is fast. You want to do the callout about `Singleton` `callout:singleton[1]`. Etc. It does not make sense that you have `@Introspected` copy duplicated in twenty files.
> 
> From the perspective of someone reviewing the guide. It is the same as common snippets.

@sdelamo The main difference I see with the current common snippets is that the existing ones are used for (almost) every guide and they contain things that we don't change: header, create an app, graalvm, next step,... 
These new snippets are very specific and without reviewing all of them you don't know they are there. 


> It is a bit weird to have all those 1-line includes, but I like that it improves consistency. I tend to look at other guides to copy/paste from for things like these, so in the future I'd just copy the macro instead of the full text, or neither and write my own inconsistent version.

I agree that it helps with making all guides consistent but to be honest, I think sometimes we "over-explain" things and these new common snippets are the proof of it. We have explained what `@Singleton` or `@MicronautTest` in the basic guides so I don't see the point of repeat that over and over in more guides. If I'm new to Micronaut and I start with an advanced guide and I don't know about some annotation I have a few options: read getting started guides, read the docs, ask on gitter.


> One thing that would help would be a list of these somewhere. I've been thinking about adding some more docs for guide authors, mostly things I discovered only by reading lots of code or in PR comments. I added #420 to get this into the queue

Yeah, I totally agree with this. We (Sergio and I) discussed this a long time ago but we never write a proper documentation about all the options, macros and tricks that you can use when writing a guide.

ilopmar commented on buildSrc/src/main/groovy/io/micronaut/guides/GuideAsciidocGenerator.groovy:
```diff
@@ -133,23 +138,63 @@ class GuideAsciidocGenerator {
-
-                rawLines.add("// Start: ${commonFileName}".toString())
-                rawLines.addAll(expandAllCommonIncludes(commonFile.readLines(), destinationFolder))
-                rawLines.add("// End: ${commonFileName}".toString())
```
I think we should keep the `// Start` and `// End` comments because they help debugging the guide generated. It's not something that we use everyday but when there is a problem with our macros it helps to see from which snippet a specific text is coming.

sdelamo replied:
I think, at most, this should be a debug setting. Not a default setting.

ilopmar reviewed (changes requested):
There is a typo in the name of `callout-configuration-properties-contraints.adoc`, a missing `s` in constraints.

After I fixed it there are a few warnings when rendering all the guides, please fix them:

```
Aug 05, 2021 9:28:04 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb next_block
WARNING: micronaut-email-gradle-groovy.adoc: line 262: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:06 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb parse
WARNING: micronaut-azure-http-functions-gradle-groovy.adoc: line 78: no callout found for <1>
Aug 05, 2021 9:28:06 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb parse
WARNING: micronaut-azure-http-functions-gradle-java.adoc: line 83: no callout found for <1>
Aug 05, 2021 9:28:08 AM  
WARNING: micronaut-azure-http-functions-gradle-kotlin.adoc: line 83: no callout found for <1>
Aug 05, 2021 9:28:08 AM  
WARNING: micronaut-email-maven-groovy.adoc: line 272: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:09 AM  
WARNING: micronaut-multitenancy-propagation-gradle-groovy.adoc: line 308: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:11 AM  
WARNING: micronaut-database-authentication-provider-gradle-groovy.adoc: line 140: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:11 AM  
WARNING: micronaut-database-authentication-provider-gradle-groovy.adoc: line 252: callout list item index: expected 2, got 0
Aug 05, 2021 9:28:11 AM  
WARNING: micronaut-azure-http-functions-maven-kotlin.adoc: line 83: no callout found for <1>
Aug 05, 2021 9:28:13 AM  
WARNING: micronaut-email-gradle-java.adoc: line 267: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:13 AM  
WARNING: micronaut-azure-http-functions-maven-groovy.adoc: line 78: no callout found for <1>
Aug 05, 2021 9:28:13 AM  
WARNING: micronaut-microservices-services-discover-eureka-maven-groovy.adoc: line 217: no callout found for <1>
Aug 05, 2021 9:28:15 AM  
WARNING: micronaut-microservices-services-discover-eureka-gradle-groovy.adoc: line 217: no callout found for <1>
Aug 05, 2021 9:28:17 AM  
WARNING: micronaut-email-maven-java.adoc: line 277: callout list item index: expected 2, got 1
Aug 05, 2021 9:28:17 AM  
WARNING: micronaut-azure-http-functions-maven-java.adoc: line 83: no callout found for <1>
```

sdelamo commented:
> I think sometimes we "over-explain" things and these new common snippets are the proof of it. We have explained what @Singleton or @MicronautTest in the basic guides so I don't see the point of repeat that over and over in more guides. If I'm new to Micronaut and I start with an advanced guide and I don't know about some annotation I have a few options: read getting started guides, read the docs, ask on gitter

I have the opposite view. I think we should explain things over and over. I don't think we should assume anyone reading a guide has previously read another guide or any other Micronaut Material. 

If I explain you something that you already know (and the explanation is one line) and you are offended/annoyed by that. Well...,  I don't think that those users should be the one defining how we write the guides. 

I think we should always err on the side of over explaining.

sdelamo committed:
add missing s to constraints file name

sdelamo committed:
revert deletion of debug statements

sdelamo committed:
fix callouts

sdelamo commented:
> There is a typo in the name of `callout-configuration-properties-contraints.adoc`, a missing `s` in constraints.
> 
> After I fixed it there are a few warnings when rendering all the guides, please fix them:
> 
> ```
> Aug 05, 2021 9:28:04 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb next_block
> WARNING: micronaut-email-gradle-groovy.adoc: line 262: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:06 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb parse
> WARNING: micronaut-azure-http-functions-gradle-groovy.adoc: line 78: no callout found for <1>
> Aug 05, 2021 9:28:06 AM uri:classloader:/gems/asciidoctor-2.0.12/lib/asciidoctor/parser.rb parse
> WARNING: micronaut-azure-http-functions-gradle-java.adoc: line 83: no callout found for <1>
> Aug 05, 2021 9:28:08 AM  
> WARNING: micronaut-azure-http-functions-gradle-kotlin.adoc: line 83: no callout found for <1>
> Aug 05, 2021 9:28:08 AM  
> WARNING: micronaut-email-maven-groovy.adoc: line 272: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:09 AM  
> WARNING: micronaut-multitenancy-propagation-gradle-groovy.adoc: line 308: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:11 AM  
> WARNING: micronaut-database-authentication-provider-gradle-groovy.adoc: line 140: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:11 AM  
> WARNING: micronaut-database-authentication-provider-gradle-groovy.adoc: line 252: callout list item index: expected 2, got 0
> Aug 05, 2021 9:28:11 AM  
> WARNING: micronaut-azure-http-functions-maven-kotlin.adoc: line 83: no callout found for <1>
> Aug 05, 2021 9:28:13 AM  
> WARNING: micronaut-email-gradle-java.adoc: line 267: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:13 AM  
> WARNING: micronaut-azure-http-functions-maven-groovy.adoc: line 78: no callout found for <1>
> Aug 05, 2021 9:28:13 AM  
> WARNING: micronaut-microservices-services-discover-eureka-maven-groovy.adoc: line 217: no callout found for <1>
> Aug 05, 2021 9:28:15 AM  
> WARNING: micronaut-microservices-services-discover-eureka-gradle-groovy.adoc: line 217: no callout found for <1>
> Aug 05, 2021 9:28:17 AM  
> WARNING: micronaut-email-maven-java.adoc: line 277: callout list item index: expected 2, got 1
> Aug 05, 2021 9:28:17 AM  
> WARNING: micronaut-azure-http-functions-maven-java.adoc: line 83: no callout found for <1>
> ```


I fixed those warnings.

ilopmar commented:
> I think we should always err on the side of over explaining.

Well... Maybe you are right and now with these common callouts snippets it would be easy to explain things all the time.

sdelamo commented:
about being able to write macros easily I have writen a blog post (GIF at the bottom)

https://sergiodelamo.com/blog/text-expander-dropdown-macros.html

ilopmar commented:
> about being able to write macros easily I have writen a blog post (GIF at the bottom)
> 
> [sergiodelamo.com/blog/text-expander-dropdown-macros.html](https://sergiodelamo.com/blog/text-expander-dropdown-macros.html)

Looks cool! 

For the ones that don't use Mac but Linux, we could do the same with IntelliJ live templates, and we could even export and share them #justSaying :sunglasses:
